packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -192,187 +191,199 @@ export function typeDefinition<T>(
+			groupNode: value.definitionsByParent[0],
 			typeDef: d,
+			children: [],
 		}))
 	}
 	const nodeQueue: CheckerTreeNode<T>[] = [rootNode]
 
 	while (nodeQueue.length !== 0) {
 		const node = nodeQueue.splice(0, 1)[0]
-		const encounterdErrors: CheckerTreeError<T>[] = []
 
 		for (const value of node.possibleValues) {
 			const inferredSimplified = simplify(value.node.inferredType, { ctx, node: value })
 			const children = ctx.getChildren(value.node.originalNode, inferredSimplified)
 			const childNodes = children.map(c => {
 				const ans: CheckerTreeNode<T> = {
 					parent: value,
-					key: !Array.isArray(c) ? { runtimeValue: c.key, typeDef: undefined } : undefined,
+					runtimeKey: !Array.isArray(c) ? c.key : undefined,
 					possibleValues: [],
 				}
 
 				ans.possibleValues = (Array.isArray(c) ? c : c.possibleValues).map(v => ({
 					entryNode: ans,
 					node: v,
-					validDefinitions: [],
+					definitionsByParent: [],
 					condensedErrors: [],
 					children: [],
 				}))
 
 				return ans
 			})
 
-			for (const def of value.validDefinitions) {
-				const { errors, childDefinitions } = checkShallowly(
-					value.node,
-					inferredSimplified,
-					children,
-					def.typeDef,
-					ctx,
-				)
-				encounterdErrors.push(...errors.map(e => ({ error: e, definitionNode: def })))
-				for (let i = 0; i < childDefinitions.length; i++) {
-					const childDef = childDefinitions[i]
-					if (!childDef) {
-						continue
-					}
-					const child = childNodes[i]
-					if (child.key) {
-						child.key.typeDef = childDef.keyType
-					}
-					for (const childValue of child.possibleValues) {
-						// TODO We need some sort of map / local cache which keeps track of the original
-						// non-simplified types and see if they have been compared yet. This is needed
-						// for structures that are cyclic, to essentially bail out once we are comparing
-						// the same types again and just collect the errors of the lower depth.
-						// This will currently lead to a stack overflow error when e.g. comparing two
-						// text component definitions
-						const simplified = simplify(childDef.type, { ctx, node: childValue })
-						// TODO this does not keep track correctly of empty unions. The child node should receive
-						// some kind of empty union valid definition with the parent set to the correct definition
-						// so that we can potentially error some valid parent defs if only some of them produce an
-						// empty union for the child.
-						const validDefs =
-							(simplified.kind === 'union' ? simplified.members : [simplified]).map(d => ({
+			for (const definitionGroup of value.definitionsByParent) {
+				const definitionErrors: ErrorCondensingDefinition<T>[] = []
+
+				if (definitionGroup.validDefinitions.length === 0) {
+					// nothing can be assigned to an empty union
+					definitionGroup.condensedErrors = [[{
+						kind: 'type_mismatch',
+						node: value.node,
+						expected: [],
```

Comment:
This is not accounted for in the error reporter

Reply:
I don't think it needs to have a special check?

I checked the implementation of `arrayToMessage` and it looks like it returns a translated `nothing` when an empty array is supplied.

I haven't tested this though.

Reply:
Oh it does? Neat! my bad